app/code/Magento/PageCache/view/frontend/web/js/prolong-cookie.js

```diff
@@ -0,0 +1,38 @@
+/**
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+define([
+    'jquery',
+    'jquery/ui'
+], function ($) {
+    'use strict';
+
+    $.widget('mage.prolongCookie', {
+        options: {
+            prolongActionUrl: '',
+        },
+
+        /**
+         * Widget constructor.
+         * @private
+         */
+        _create: function () {
+            this._prolongCookie();
+        },
+
+        /**
+         * Prolongs cookie(s) lifetime by performing post ajax request.
+         * @private
+         */
+        _prolongCookie: function () {
+            if (this.options.prolongActionUrl.length === 0) {
+                return;
+            }
+
+            $.post(this.options.prolongActionUrl);
```

Comment:
I don't think that is a good idea to add one more request on every page. We should send post request only in x  second before session end.

Reply:
Thanks for your comment.

I thought about this too. But in the case Varnish is used as caching application, we may not reach Backend at all during configured cookie lifetime. So we may increase configured cookie lifetime up to 2 times by sending such request just before session is exceeded, for instance, when customer is reading the same page all this time. I guess such behavior is not expected.

There is one more idea that just has come to my mind, but it's more tricky. We can introduce another cookie to store expected lifetime of frontend cookie. It may be updated by JS every time the customer opens new page, even fully cached. And when the expected lifetime of frontend cookie will be much greater than its real lifetime in browser, we will send such prolongation request and synchronize the lifetime of frontend cookie (by calculated delta).

There are some other nuances that should be kept in mind, but what do you think in general, should we implement a tricky logic in this case?

Reply:
I believe that making a request on every page (or only on cached) is not acceptable solution compared with a workaround like significant increase of session/cookies or switch to built-in FPC. So we need more advanced/tricky implementation. 
Also cookies expiration data can be added to private data. See http://devdocs.magento.com/guides/v2.2/extension-dev-guide/cache/page-caching/private-content.html for more details

Reply:
Agreed, will come back with updated solution soon.
Thanks for the advice regarding private data.

Reply:
Ok, thank you for your cooperation